lighthouse-core/scripts/generate-bundlephobia-database.js

```diff
@@ -0,0 +1,145 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+/* eslint-disable no-console */
+
+// TODO(saavan) These type definitions are incorrect and need to be fixed.
+/** @typedef {{gzip: number, name: string, version: string | number, repository: string}} BundlePhobiaLibrary */
+
+const fs = require('fs');
+const getPackageVersionList = require('bundle-phobia-cli').fetchPackageStats.getPackageVersionList;
+const fetchPackageStats = require('bundle-phobia-cli').fetchPackageStats.fetchPackageStats;
+
+const databasePath = '../audits/byte-efficiency/bundlephobia-database.json';
+
+/** @type {Record<string, string[]>} */
+const suggestionsJSON = require('../audits/byte-efficiency/library-suggestions.js').suggestions;
+/** @type string[] */
+/* eslint-disable-next-line max-len */
+const librarySuggestions = Object.keys(suggestionsJSON).map(s => suggestionsJSON[s].concat(s)).flat();
+
+/** @type {Record<string, Record<'lastScraped', number|string> | Record<'repository', string> | Record<string, BundlePhobiaLibrary>>} */
+let database = {};
+if (fs.existsSync(databasePath)) {
+  database = require(databasePath);
+}
+
+/**
+ * Returns true if this library has been scraped from BundlePhobia in the past week.
+ * @param {string} library
+ * @return {boolean}
+ */
+function hasBeenRecentlyScraped(library) {
+  if (!database[library] || database[library].lastScraped === 'Error') return false;
+  return (Date.now() - database[library].lastScraped) / (1000 * 60 * 60 * 24.0) < 7;
+}
+
+/**
+ * Returns true if the object represents valid BundlePhobia JSON.
+ * The version string must not match this false-positive expression: '{number} packages'.
+ * @param {any} library
+ * @return {boolean}
```

Comment:
typescript lets you do some cool stuff like

```suggestion
 * @return {library is import('bundle-phobia-cli').BundlePhobiaLibrary}
```

which then lets you use it as that type

for example

```js
/** @typedef {{type: 'foo', foo: number}} Foo */

/** 
 * @param {unknown} o 
 * @return {o is Foo} 
 */
function isFoo(o) {
    return typeof o === 'object' && o && 'type' in o && 'foo' in o && o.type === 'foo' && typeof o.foo === 'number'
}

/** @type {unknown} */
const thing = {};
if (isFoo(thing)) console.log(thing.foo) // not a type error
else console.log(thing.foo) // type error
```


all that being said, I don't think you really need this here. You've already added a typedef to the CLI export so we can already trust it's an object like that. if it sometimes returns bad data, maybe we need to change the CLI typedef to reflect that?

Reply:
Ohh gotcha. So the function could theoretically network request fail which is why I think we should keep the validation function. Also it helps filter out false-positive results where the `version` attribute is like `{number} packages`.

Also I'm getting an `invalid type syntax` error

![image](https://user-images.githubusercontent.com/66381097/88439923-21ac7d00-cdd2-11ea-9e81-7c4b5db960ba.png)

Reply:
Oh sorry you might need to pull out `import('bundle-phobia-cli').BundlePhobiaLibrary` into a typedef I've never tried it all in oneline before 😆 

`/** @typedef {import('bundle-phobia-cli').BundlePhobiaLibrary} BundlePhobiaLibraryFromAPI */

Reply:
> So the function could theoretically network request fail which is why I think we should keep the validation function.

If it fails a network request it doesn't throw an error but returns some unknown object? Hm, that's an odd API choice, then let's reflect that in the typedef of the module.

Reply:
> Oh sorry you might need to pull out import('bundle-phobia-cli').BundlePhobiaLibrary into a typedef I've never tried it all in oneline before 😆

I tried that too but I'm still getting a syntax highlighting error rip


> If it fails a network request it doesn't throw an error but returns some unknown object? Hm, that's an odd API choice, then let's reflect that in the typedef of the module.

Should I make each attribute optional?

```
interface BundlePhobiaLibrary {
    size?: number;
    version?: string;
    etc
  }
```

Reply:
> I tried that too but I'm still getting a syntax highlighting error rip

Something might be up with your editor? This works exactly as intended in our repo, any difference in what you're trying syntax-wise?

![image](https://user-images.githubusercontent.com/2301202/89046694-f6b2b380-d312-11ea-974d-b8ed43e77384.png)

> Should I make each attribute optional?

I think we may have missed a step here :) You mentioned that you were doing this checking because the bundle phobia API returns to you a different object when the network fails. If that's the case, then we should reflect that in the typedef by returning whatever type that different object is.

```typescript
// This one should be if everything works as expected
interface BundlePhobiaLibrary {
  size: number
  version: string
  // ...
}

interface BundlePhobiaError {
  // ?? I don't know what this looks like, but if you found it returning something else it should match that
}


export namespace fetchPackageStats {
  export function getPackageVersionList(packageName: string): string[];
  export function fetchPackageStats(packageName: string): Array<BundlePhobiaLibrary|BundlePhobiaError>;
}
```

If the network error results in the object being completely empty with no keys, then yes, it could make sense to say that the return type of `fetchPackageStats` is just `Partial<BundlePhobiaLibrary>`